Unify nf-lang config scopes with runtime classes

bentsherman opened the pull request:
This PR moves the config schema annotations into the runtime so that we don't have to update config options in both places.

Plugin config scopes are now defined in the plugins themselves. Core plugins are typically only loaded when they are actually used (i.e. the awsbatch executor), so when Nextflow validates config at runtime, it will skip validation of core plugin config if the core plugin isn't loaded. This allows the `lint` command to work the same as before.

I will update the language server to extract this metadata at build-time with a custom Gradle task. For third-party plugins we will need to publish this metadata to the plugin registry, but that can be done later.

I have removed several config options that were undocumented or appeared to be legacy options. We can preserve them if needed. See my comments below.

TODO:
- [x] Update tests
- [x] Publish JSON artifact for language server

christopher-hakkaart reviewed (commented):
Docs look good. No comments.

bentsherman commented on modules/nextflow/src/main/groovy/nextflow/container/ContainerConfig.groovy:
```diff
@@ -17,129 +17,51 @@
 
-    boolean isEnabled() {
-        return get('enabled')?.toString() == 'true'
+    default boolean entrypointOverride() {
+        return ContainerHelper.entrypointOverride()
     }
```
The `entrypointOverride` is a hidden config option that can be set for `docker`, `k8s`, and the other container config scopes.

I made it so that it can only be enabled by setting the `NXF_CONTAINER_ENTRYPOINT_OVERRIDE` environment variable, since that seems simpler than adding a config option for every container scope.

If we want to preserve this behavior, I suggest we document this env var and add a note to the 25.10 migration notes.

bentsherman replied:
Related: https://github.com/nextflow-io/nextflow/commit/b3a4bf857cef0a8b16814baf9e13aa9296ca208a

bentsherman commented on modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy:
```diff
@@ -728,22 +733,17 @@ class BashWrapperBuilder {
+            builder.params(kill: containerConfig.getKill())
 
-        if( containerConfig.writableInputMounts==false )
+        if( containerConfig.writableInputMounts()==false )
             builder.params(readOnlyInputs: true)
```
It looks like some container scopes can specify `writableInputMounts = false` to mount folders into the container as read-only. There are tests for this option for apptainer, charliecloud, docker, podman, and singularity... so basically all of them.

If we want to keep this behavior, I suggest we add config option for each container scope where it's relevant. Or we could add an environment var.

bentsherman replied:
I think I will add explicit config options for `writableInputMounts` just for docker and charliecloud for now, and if people request it for other containers we can add it

bentsherman commented on plugins/nf-k8s/src/main/nextflow/k8s/K8sConfig.groovy:
```diff
@@ -89,43 +248,30 @@ class K8sConfig implements Map<String,Object> {
     }
 
     boolean getCleanup(boolean defValue=true) {
-        target.cleanup == null ? defValue : Boolean.valueOf( target.cleanup as String )
+        defValue
     }
```
There is a hidden option `k8s.cleanup` which is enabled by default, but can be disabled to not delete any pods. Normally only successful pods are deleted. Not sure if we still need this setting.

bentsherman commented on plugins/nf-k8s/src/main/nextflow/k8s/K8sConfig.groovy:
```diff
@@ -140,74 +286,36 @@ class K8sConfig implements Map<String,Object> {
-    }
-
-    boolean getAutoMountHostPaths() {
-        Boolean.valueOf( target.autoMountHostPaths as String )
+        return "nextflow/nextflow:${BuildInfo.version}"
     }
```
There is a hidden option `k8s.nextflow.image` which controls the nextflow container image used by `kuberun`. This can already be controlled via CLI option so I don't think we need a config option for it.

bentsherman commented on plugins/nf-k8s/src/main/nextflow/k8s/K8sConfig.groovy:
```diff
@@ -217,50 +325,27 @@ class K8sConfig implements Map<String,Object> {
-
-        if( target.httpConnectTimeout )
-            result.httpConnectTimeout = target.httpConnectTimeout as Duration
+        final result = clientDiscovery(context, namespace, serviceAccount)
```
There is a hidden scope `k8s.client` which can be used to define settings for the K8s client directly in Nextflow config. I think the best practice is to define these settings in `.kube/config`. If we want to keep this approach we should explicitly document it.

jorgee commented on modules/nf-lineage/src/main/nextflow/lineage/LinStoreFactory.groovy:
```diff
@@ -51,15 +51,14 @@ abstract class LinStoreFactory implements ExtensionPoint {
             initialized = true
-            final config = LineageConfig.create(session)
-            if( !config.enabled )
+            if( requireConfig && !config.enabled )
```
I think this statement is not correct, if lineage is not enabled, it should return null for whatever `requireConfig` value.  When running the commands, requireConfig is false, and it always creates the config.

jorgee commented on modules/nf-lineage/src/main/nextflow/lineage/cli/LinCommandImpl.groovy:
```diff
@@ -193,8 +192,8 @@ class LinCommandImpl implements CmdLineage.LinCommand {
     @Override
-    void check(ConfigMap config, List<String> args) {
-        final store = LinStoreFactory.getOrCreate(new Session(config))
+    void check(Map config, List<String> args) {
+        final store = LinStoreFactory.getOrCreate(LineageConfig.create(config), false)
```
I was wrong, the error of the second test (`should print correct validate path`) is related to the session.  The filesystem is getting the configuration from the session, and this is why it is  using the default location. 

I have tried to replace the `create` call in the `LidFileSystem` constructor by `getOrCreate` but then several Lin FS tests are failing because they are reusing the store. I should modify them to reset the store at the end of each test. I think it is worth to do in another PR.

bentsherman commented:
Tests are passing now and I added a page to the developer docs

I also added the JSON and Markdown renderers. They aren't used by Nextflow itself. The JSON renderer is used by the language server, but it will also be needed by plugins in the future so it makes sense to keep the core logic in Nextflow.

pditommaso commented:
Wonder why the config classes need to be specified in the plugin extension file

bentsherman commented:
Because they extend `ConfigScope` which is an extension point. Then `ConfigValidator` uses the plugin system to find all config scopes

bentsherman commented:
I'm fine with using the ServiceLoader if it's something you want to apply to the rest of the codebase, otherwise I would rather stick to the existing system. Maybe the ServiceLoader approach should be investigated separately

pditommaso commented:
let's disable that test(s) temporary

pditommaso reviewed (approved):
Great job! let's go
